plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilter.java

```diff
@@ -137,4 +142,106 @@ public void failed(Exception e) {
     public int order() {
         return Integer.MAX_VALUE;
     }
+
+    final static class BulkProcessingState implements Iterator<ActionRequest> {
+
+        final String action;
+        final String pipelineId;
+        final ActionFilterChain chain;
+        final BulkRequest bulkRequest;
+        final ActionListener<BulkResponse> listener;
+
+        final Set<Integer> failedSlots;
+        final List<BulkItemResponse> itemResponses;
+
+        int currentSlot = -1;
+
+        BulkProcessingState(String action, String pipelineId, ActionFilterChain chain, BulkRequest bulkRequest,
+                            ActionListener<BulkResponse> listener) {
+            this.action = action;
+            this.pipelineId = pipelineId;
+            this.chain = chain;
+            this.bulkRequest = bulkRequest;
+            this.listener = listener;
+            this.failedSlots = new HashSet<>();
+            this.itemResponses = new ArrayList<>(bulkRequest.requests().size());
+        }
+
+        @Override
+        public ActionRequest next() {
+            return bulkRequest.requests().get(++currentSlot);
+        }
+
+        @Override
+        public boolean hasNext() {
+            return (currentSlot + 1) < bulkRequest.requests().size();
+        }
+
+        void proceed() {
+            if (itemResponses.isEmpty()) {
+                chain.proceed(action, bulkRequest, listener);
+            } else {
+                BulkRequest modifiedBulkRequest = new BulkRequest(bulkRequest);
+                modifiedBulkRequest.refresh(bulkRequest.refresh());
+                modifiedBulkRequest.consistencyLevel(bulkRequest.consistencyLevel());
+                modifiedBulkRequest.timeout(bulkRequest.timeout());
+
+                List<Integer> originalSlots = new ArrayList<>();
+                for (int i = 0; i < bulkRequest.requests().size(); i++) {
+                    ActionRequest request = bulkRequest.requests().get(i);
+                    if (failedSlots.contains(i) == false) {
+                        originalSlots.add(i);
+                        modifiedBulkRequest.add(request);
+                    }
+                }
+
+                if (modifiedBulkRequest.requests().isEmpty()) {
+                    // in this stage, the transport bulk action can't deal with a bulk request with no requests,
+                    // so we stop and send a empty response back to the client.
+                    // (this will happen if all preprocessing all items in the bulk failed)
+                    listener.onResponse(new BulkResponse(itemResponses.toArray(new BulkItemResponse[0]), 0));
+                } else {
+                    chain.proceed(action, modifiedBulkRequest, new IngestBulkResponseListener(originalSlots, itemResponses, listener));
+                }
+
+            }
+        }
+
+        void markBulkItemAsFailed(IndexRequest indexRequest, Exception e) {
```

Comment:
can we call it `markCurrentItemAsFailed` ? Also can we not retrieve from the bulkrequest given currentSlot the index request rather than requiring it as an argument?